File: src/PlaneFlyer.java
import mmm.*;
import lejos.nxt.*;

/**
 * MMM Module Plane Flyers from the video game Theme Park.
 * Building instructions for MMM modules are on brickhub.org
 * @author Lasse Deleuran
 */
public class PlaneFlyer {
	private static final Track track = new Track(MotorPort.C, true);
	private static final NXTRegulatedMotor lifter = new NXTRegulatedMotor(MotorPort.A);
	private static final NXTRegulatedMotor turner = new NXTRegulatedMotor(MotorPort.B);
	private static final NXTMotor turnResetter = new NXTMotor(MotorPort.B);
	
	private static final FigureSensor figureSensor = new FigureSensor(SensorPort.S3);
	private static final TouchSensor touch1 = new TouchSensor(SensorPort.S4);
	private static final TouchSensor touch2 = new TouchSensor(SensorPort.S2);

	public static final int CAPACITY = 4;

	// Speed and acceleration:
	public static final int ACCELERATION = 300;	
	public static final int SPEED_TURN_SLOW = 150;
	public static final int SPEED_TURN_FAST = 400;
	public static final int SPEED_LIFT = 230;

	// Movements:
	public static final int LIFT_CLEAR = -300;
	public static final int LIFT_TOP = -250;
	
	public static final int TURN_DOOR = -120;
	
	private static int guests = 0;
	
	public static void main(String[] args) {
		init();
		
		while(true) {
			for(int i = 0; i < CAPACITY && pickup(); i++)
				;
			if(guests > 0) {
				lifter.rotate(LIFT_TOP);
				turner.setSpeed(SPEED_TURN_FAST);
				turn(3 * 360 - (90 + 90*guests));
				turner.setSpeed(SPEED_TURN_SLOW);
				lifter.rotate(-LIFT_TOP);
				while(guests > 0) {
					leave();
				}
			}
			Time.sleep(15*1000); // Wait 15 seconds until next round
		}
	}
	
	/**
	 * Attempt to pick up a guest.
	 * At return, a plane is over the entrance of the track.
	 * This plane might be empty if no passenger has arrived.
	 * @return true if another guest should be picked up.
	 */
	private static boolean pickup() {		
		turner.setSpeed(SPEED_TURN_SLOW); // Slow for pickups.
		track.in();
		track.boost(170); // Change direction

		// Wait for guest. If no guest arrives, then return false:
		if(!figureSensor.seesMinifig(20*1000)) {
			track.resetSpeed();
			track.out();
			
			if(guests > 0) { // Only close airplane if there are guests:
				liftToDoor();
				closeDoor();
			}
			return false;
		}
		
		// Get passenger into plane:
		track.boost(375);
		guests++;
		
		// Close the door:
		track.stop();
		liftToDoor();		
		track.out();
		closeDoor();

		if(guests < CAPACITY) {
			// Position next plane:
			openDoor();
			downAndReset();
			return true;
		}
		else {			
			return false;
		}
	}
	
	private static void leave() {
		turner.setSpeed(SPEED_TURN_SLOW); // Slow for pickups.

		openDoor();
		
		downAndReset();
		
		clear();
		guests--;

		if(guests > 0) {
			liftToDoor();
			closeDoor();
		}
	}
	
	private static void clear() {
		track.out();
		int tries = 0;
		Time.sleep(4500);		
		while(figureSensor.seesMinifig(2000)) {
			// Try to get minifig out if stuck:
			Time.sleep(3000);
			track.in();
			Time.sleep(900);
			track.out();
			Time.sleep(3000);
			tries++;
			if(tries == 2)
				track.boost(); // People can't get out!
		}
		track.resetSpeed();
	}
	
	private static boolean recentlyAdjusted = false;
	private static void init() {
		track.out();
		
		Button.RIGHT.addButtonListener(new ButtonListener() {			
			@Override
			public void buttonReleased(Button b) {
				// NOP
			}
			
			@Override
			public void buttonPressed(Button b) {
				recentlyAdjusted = true;
				turner.rotate(-10);
			}
		});
		
		Button.LEFT.addButtonListener(new ButtonListener() {			
			@Override
			public void buttonReleased(Button b) {
				// NOP
			}
			
			@Override
			public void buttonPressed(Button b) {
				recentlyAdjusted = true;
				turner.rotate(10);
			}
		});
		
		figureSensor.resume();
		do {
			recentlyAdjusted = false;
			Time.sleep(5000);
		}
		while(recentlyAdjusted);
		
		lifter.setSpeed(SPEED_LIFT);
		if(touch1.isPressed() || touch2.isPressed()) {
			lifter.backward();
			while(touch1.isPressed() || touch2.isPressed())
				;
		}
		downAndReset();
		KillSwitch.enable();
		
		// Assume turn motor is placed with a plane over the entrance:
		turner.setAcceleration(ACCELERATION);
		turner.flt();
		
		// Set track:
		track.in();
	}
	
	private static void turn(float planeAngle) {
		planeAngle = -planeAngle; // Turn counter clockwise, as that is forward for planes.
		//public static final int TURN_QUARTER = -450; // 90*60/12
		lifter.setSpeed(turner.getSpeed()*12/60);
		lifter.rotate((int)planeAngle, true);
		turner.rotate((int)(planeAngle*60/12), true);

		while(lifter.isMoving() || turner.isMoving())
			;
		
		lifter.setSpeed(SPEED_LIFT);
	}
	
	private static void liftToDoor() {
		lifter.backward();
		while(touch1.isPressed())
			;
		lifter.stop();
		lifter.rotate(-525);
	}
	
	private static void closeDoor() {
		turner.rotate(-TURN_DOOR);		
		// Clear the pin:
		lifter.rotate(LIFT_CLEAR);
		turner.rotate(TURN_DOOR);
	}
	
	private static void openDoor() {
		turn(70);
		lifter.rotate(-LIFT_CLEAR+15);
		turn(22.5f); // Imprecise NXT motor. 20 would be correct.
	}
	
	/*¨
	 * Ensure lifter is reset, so that it does not get out of calibration.
	 */
	private static void downAndReset() {		
		lifter.forward();
		while(!(touch1.isPressed() || touch2.isPressed()))
			;
		lifter.stop();
		lifter.rotate(140);
		
		// Reset turner:
		turner.flt();
		turner.suspendRegulation();
		
		turnResetter.setPower(15);
		turnResetter.forward();
		Time.sleep(170);
		turnResetter.backward();
		Time.sleep(170);
		turnResetter.forward();
		Time.sleep(170);
		turnResetter.flt();
		
		// Restore turner:
		turner.setSpeed(SPEED_LIFT);
		turner.resetTachoCount();
	}
}

File: src/PlaneFlyer.py
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor, InfraredSensor, ForceSensor
from pybricks.parameters import Button, Color, Direction, Port, Stop
from pybricks.tools import wait, StopWatch

hub = InventorHub()
watch = StopWatch()

track = Motor(Port.E, Direction.COUNTERCLOCKWISE)
innerTrack = Motor(Port.C)
lifter = Motor(Port.B)
turner = Motor(Port.F)

figureSensor = InfraredSensor(Port.A)
heightSensor = ForceSensor(Port.D)

CAPACITY = 4

# Speed:
SPEED_TURN_ADJUST = 80
SPEED_TURN_SLOW = 110
SPEED_TURN_FAST = 450
SPEED_LIFT = 230
SPEED_TRACK = 130
SPEED_INNER_TRACK = 400

# Movements:
FORCE_LOAD = 4.89
FORCE_DOOR_CLOSE = 3.35
FORCE_PIN = 2.992
FORCE_CLEAR = 1.5

TURN_DOOR = 35

# State:
guests = 0

def turn(planeAngle, speed_turner = SPEED_TURN_FAST):
    print('Turning', planeAngle)
    speed_lifter = speed_turner*12/60
    lifter.run_angle(speed_lifter, -planeAngle, Stop.HOLD, wait=False)
    turner.run_angle(speed_turner, planeAngle*60/12, Stop.HOLD, wait=False)
    while not (lifter.done() and turner.done()):
        wait(100)

def liftTo(force):
    watch.reset()
    f = heightSensor.force()
    print('lift', f, '->', force)
    #    raise SystemExit()
    if heightSensor.force() > force:
        lifter.run(-SPEED_LIFT)
        while heightSensor.force() > force:
            wait(50)
    else:
        lifter.run(SPEED_LIFT)
        while heightSensor.force() < force:
            wait(50)
            while(watch.time() > 8000):
                lifter.stop()
                print('DEAD')
                wait(100000) # Dead
    lifter.stop()

def turnToSensor():
    print('Turn to sensor')
    turner.run(SPEED_TURN_ADJUST)
    while heightSensor.force() == 0.0:
        wait(50)
    turner.stop()
    print('Turned to force', heightSensor.force())

def openDoor():
    print('Opening door')
    liftTo(FORCE_CLEAR)
    turn(-TURN_DOOR, SPEED_TURN_SLOW)
    lifter.run_angle(SPEED_LIFT, 230) # Down over pin.
    turn(TURN_DOOR, SPEED_TURN_SLOW)
    liftTo(FORCE_LOAD)

def closeDoor():
    print('Closing door')
    liftTo(FORCE_DOOR_CLOSE)
    innerTrack.run(SPEED_TRACK) # Clear stragglers
    # Close the door:
    turner.run(-SPEED_TURN_SLOW)
    while heightSensor.force() > 0.0:
        wait(50)
    turner.stop()
    lifter.run_angle(SPEED_LIFT, -450) # Clear the pin
    turnToSensor()
    turn(18, SPEED_TURN_SLOW)
    innerTrack.stop()

def init():
    global guests
    f = heightSensor.force()
    print('init', f)
    if f >= FORCE_LOAD:
        print('Already above sensor')
        guests = 0
        return
    guests = 4 # Unknown number of guests!
    if f > 0.0:
        print('Lift clear for init')
        liftTo(0.0)
        turn(50, SPEED_TURN_SLOW)
    while heightSensor.force() > 0.0:
        turn(20, SPEED_TURN_SLOW)
    turnToSensor()
    print('Adjusting under new plane', heightSensor.force())
    turn(18, SPEED_TURN_SLOW)
    liftTo(FORCE_LOAD)
    while guests > 0:
        leave()

def pickup():
    global guests
    track.run(SPEED_TRACK)
    # Wait for guest:
    ticksFree = 0
    while ticksFree <= 5: # Wait until sensor is free for some time:
        wait(100)
        if figureSensor.distance() < 19:
            ticksFree = 0
        else:
            ticksFree = ticksFree+1
    cur = figureSensor.distance()
    prev = cur
    while cur <= prev or cur > 10: # Wait until figure is close to sensor:
        prev = cur
        wait(50)
        cur = figureSensor.distance()
    
    # Get passenger into plane:
    track.stop()
    innerTrack.run(-SPEED_INNER_TRACK)
    track.run_angle(SPEED_TRACK, 200)
    innerTrack.stop()
    innerTrack.run_angle(-SPEED_INNER_TRACK, 1300)
    track.run(SPEED_TRACK)
    guests = guests + 1
    # Close the door:
    closeDoor()
    if guests < CAPACITY: # Position next plane:
        turn(90, SPEED_TURN_SLOW)
        openDoor()

def leave():
    global guests
    openDoor()
    innerTrack.run_angle(SPEED_INNER_TRACK, 400)
    timeFree = 0
    while timeFree < 3000:
        if figureSensor.distance() < 19:
            timeFree = 0 
        wait(50)
        timeFree = timeFree + 50
    innerTrack.run_angle(SPEED_INNER_TRACK, 1000)
    guests = guests - 1
    if guests > 0:
        closeDoor()
        lifter.run_angle(SPEED_LIFT, -200)
        turn(90, SPEED_TURN_SLOW)

#pickup()
#init()
#closeDoor()
#openDoor()
#raise SystemExit()

# Main loop:
print('Plane Flyer Debugging Enabled', figureSensor.distance())
track.run(SPEED_TRACK)
init()
while True:
    while guests < CAPACITY:
        pickup()
    liftTo(0.0)
    turner.control.limits(1000, 120, 560)
    lifter.control.limits(1000, 50, 560)
    turn(4 * 360, SPEED_TURN_FAST)
    turner.control.limits(1000, 2000, 560) # Read using print(turner.control.limits())
    lifter.control.limits(1000, 2000, 560)
    while guests > 0:
        leave()
